clippy_lints/src/danger_not_accepted.rs

```diff
@@ -0,0 +1,375 @@
+        if let ExprKind::MethodCall(_path, _, _self_arg, ..) = &expr.kind
+            && let Some(fn_id) = cx.typeck_results().type_dependent_def_id(expr.hir_id)
+            // And that function is dangerous to us...
+            && let Some(dangers) = self.get_unaccepted_dangers(cx, fn_id)
+        {
+            // Raise a lint
+            emit_dangerous_call_lint(cx, expr, &dangers);
+            return;
+        }
+
+        // If we're referencing a function...
+        if let ExprKind::Path(path) = &expr.kind
+            && let QPath::Resolved(_, path) = path
+            && let def::Res::Def(def::DefKind::Fn, fn_id) = path.res
+            // And that function is dangerous to us...
+            && let Some(dangers) = self.get_unaccepted_dangers(cx, fn_id)
+        {
+            // Raise a lint
+            emit_dangerous_call_lint(cx, expr, &dangers);
+            return;
+        }
+    }
+
+    fn enter_lint_attrs(&mut self, cx: &LateContext<'_>, attrs: &'_ [rustc_ast::Attribute]) {
+        // Both `accept_danger` and `dangerous` contribute to the accepted danger map.
+        for attr_name in ["accept_danger", "dangerous"] {
+            for attr in get_attr(cx.sess(), attrs, attr_name) {
+                // We don't really care about why a user might have accepted a danger or marked a
+                // section as dangerous.
+                let (dangers, _ignored_reason) = parse_dangers_attr(cx, attr);
+
+                for (_span, danger) in dangers {
+                    *self.accepted_dangers.entry(danger).or_default() += 1;
+                }
+            }
+        }
+    }
+
+    fn exit_lint_attrs(&mut self, cx: &LateContext<'_>, attrs: &'_ [rustc_ast::Attribute]) {
+        // Both `accept_danger` and `dangerous` contribute to the accepted danger map.
+        for attr_name in ["accept_danger", "dangerous"] {
+            for attr in get_attr(cx.sess(), attrs, attr_name) {
+                let (dangers, _ignored_reason) = parse_dangers_attr(cx, attr);
+                for (_span, danger) in dangers {
+                    match self.accepted_dangers.entry(danger) {
+                        StdEntry::Occupied(mut entry) => {
+                            *entry.get_mut() -= 1;
+                            if *entry.get() == 0 {
+                                entry.remove();
+                            }
+                        },
+                        StdEntry::Vacant(_) => unreachable!(),
+                    }
+                }
+            }
+        }
+    }
+}
+
+struct UnacceptedDanger {
+    span: Span,
+    id: Symbol,
+    reason: Option<token::Lit>,
+}
+
+impl DangerNotAccepted {
+    fn get_unaccepted_dangers(&self, cx: &LateContext<'_>, item_id: def_id::DefId) -> Option<Vec<UnacceptedDanger>> {
+        let mut unaccepted_dangers = Vec::new();
+        let mut item_iter = Some(item_id);
+
+        while let Some(item_id) = item_iter {
+            item_iter = cx.tcx.opt_parent(item_id);
```

Comment:
What do you think about only supporting the danger definitions directly on functions? To me it feels similar to `unsafe`/`#[must_use]` which also apply individually

Reply:
I think, for dangers like `unchecked_foo_api`, users are likely going to reuse that same danger for many different functions (or even entire modules!) and would likely benefit from things like this:

```rust
#[clippy::dangerous(unchecked_foo_api)]
impl Foo {
    pub fn get_raw(&self) {}

    pub fn set_raw(&self) {}

    pub fn update_raw(&self) {}
}
```

Would the benefit to this be mainly a performance one? If so, there's always the possibility of applying path compression to reduce the amount of looking one may have to do.

Reply:
The perf aspect is the larger part of the concern yeah, though I haven't ran any kind of benchmark to see how much of an issue it actually is